layers/dhcpv4.go

```diff
@@ -0,0 +1,407 @@
+	DHCP_OPT_EXTENSIONS_PATH:          "OPT_EXTENSIONS_PATH",
+	DHCP_OPT_IP_FORWARDING:            "OPT_IP_FORWARDING",
+	DHCP_OPT_SOURCE_ROUTING:           "OPT_SOURCE_ROUTING",
+	DHCP_OPT_POLICY_FILTER:            "OPT_POLICY_FILTER",
+	DHCP_OPT_DGRAM_MTU:                "OPT_DGRAM_MTU",
+	DHCP_OPT_DEFAULT_TTL:              "OPT_DEFAULT_TTL",
+	DHCP_OPT_PATH_MTU_AGING_TIMEOUT:   "OPT_PATH_MTU_AGING_TIMEOUT",
+	DHCP_OPT_PATH_PLATEU_TABLE_OPTION: "OPT_PATH_PLATEU_TABLE_OPTION",
+	DHCP_OPT_INTERFACE_MTU:            "OPT_INTERFACE_MTU",
+	DHCP_OPT_ALL_SUBS_LOCAL:           "OPT_ALL_SUBS_LOCAL",
+	DHCP_OPT_BROADCAST_ADDR:           "OPT_BROADCAST_ADDR",
+	DHCP_OPT_MASK_DISCOVERY:           "OPT_MASK_DISCOVERY",
+	DHCP_OPT_MASK_SUPPLIER:            "OPT_MASK_SUPPLIER",
+	DHCP_OPT_ROUTER_DISCOVERY:         "OPT_ROUTER_DISCOVERY",
+	DHCP_OPT_ROUTER_SOLICIT_ADDR:      "OPT_ROUTER_SOLICIT_ADDR",
+	DHCP_OPT_STATIC_ROUTE:             "OPT_STATIC_ROUTE",
+	DHCP_OPT_ARP_TRAILERS:             "OPT_ARP_TRAILERS",
+	DHCP_OPT_ARP_TIMEOUT:              "OPT_ARP_TIMEOUT",
+	DHCP_OPT_ETHERNET_ENCAP:           "OPT_ETHERNET_ENCAP",
+	DHCP_OPT_TCP_TTL:                  "OPT_TCP_TTL",
+	DHCP_OPT_TCP_KEEPALIVE_INT:        "OPT_TCP_KEEPALIVE_INT",
+	DHCP_OPT_TCP_KEEPALIVE_GARBAGE:    "OPT_TCP_KEEPALIVE_GARBAGE",
+	DHCP_OPT_NIS_DOMAIN:               "OPT_NIS_DOMAIN",
+	DHCP_OPT_NIS_SERVERS:              "OPT_NIS_SERVERS",
+	DHCP_OPT_NTP_SERVERS:              "OPT_NTP_SERVERS",
+	DHCP_OPT_VENDOR_OPT:               "OPT_VENDOR_OPT",
+	DHCP_OPT_NETBIOS_IPNS:             "OPT_NETBIOS_IPNS",
+	DHCP_OPT_NETBIOS_DDS:              "OPT_NETBIOS_DDS",
+	DHCP_OPT_NETBIOS_NODE_TYPE:        "OPT_NETBIOS_NODE_TYPE",
+	DHCP_OPT_NETBIOS_SCOPE:            "OPT_NETBIOS_SCOPE",
+	DHCP_OPT_X_FONT_SERVER:            "OPT_X_FONT_SERVER",
+	DHCP_OPT_X_DISPLAY_MANAGER:        "OPT_X_DISPLAY_MANAGER",
+	DHCP_OPT_END:                      "(end)",
+	DHCP_OPT_SIP_SERVERS:              "SipServers",
+	DHCP_OPT_REQUEST_IP:               "RequestIP",
+	DHCP_OPT_LEASE_TIME:               "LeaseTime",
+	DHCP_OPT_EXT_OPTS:                 "ExtOpts",
+	DHCP_OPT_MESSAGE_TYPE:             "MessageType",
+	DHCP_OPT_SERVER_ID:                "ServerID",
+	DHCP_OPT_PARAMS_REQUEST:           "ParamsRequest",
+	DHCP_OPT_MESSAGE:                  "Message",
+	DHCP_OPT_MAX_DHCP_SIZE:            "MaxDHCPSize",
+	DHCP_OPT_T1:                       "Timer1",
+	DHCP_OPT_T2:                       "Timer2",
+	DHCP_OPT_CLASS_ID:                 "ClassID",
+	DHCP_OPT_CLIENT_ID:                "ClientID",
+}
+
+type DHCPOption struct {
+	Type   uint8
+	Length uint8
+	Data   []byte
+}
+
+func (o DHCPOption) String() string {
+	return fmt.Sprintf("Option(%v:%v)", DHCPOptionTypeStrings[o.Type], o.Data)
+}
+
+func NewDHCPOption(t uint8, data []byte) DHCPOption {
+	o := DHCPOption{Type: t}
+	if data != nil {
+		o.Data = data
+		o.Length = uint8(len(data))
+	}
+	return o
+}
+
+func (o *DHCPOption) Marshal() ([]byte, error) {
```

Comment:
When marshaling, please take in a https://godoc.org/github.com/google/gopacket#SerializeBuffer and serialize to it.  This avoids having to alloc a new byte array for every DHCP option.